cmake/SetupCodeCoverageReports.cmake

```diff
@@ -45,47 +47,71 @@ endif()
-#   Pass them in list form, e.g.: "-j;2" for -j 2
-function(add_code_coverage_target _targetname _testrunner)
+##------------------------------------------------------------------------------
+## blt_add_code_coverage_target( NAME              <Created Target Name>
+##                               RUNNER            <The command to run the tests>
+##                               SOURCE_DIRECTORIES [dir1 [dir2 ...]])
+##
+## Creates a new target with the given NAME that generates a code coverage report.
+##------------------------------------------------------------------------------
+function(blt_add_code_coverage_target)
+
+    set(options)
+    set(singleValueArgs NAME)
+    set(multiValueArgs RUNNER SOURCE_DIRECTORIES)
+
+    # parse the arguments
+    cmake_parse_arguments(arg
+        "${options}" "${singleValueArgs}" "${multiValueArgs}" ${ARGN} )
+
+    # Check/Set required parameters
+    if(NOT DEFINED arg_NAME)
+        message(FATAL_ERROR "blt_add_code_coverage_target requires a NAME parameter")
+    endif()
+
+    if(NOT DEFINED arg_RUNNER)
+        message(FATAL_ERROR "blt_add_code_coverage_target requires a RUNNER parameter")
+    endif()
+
+    set(_coverage_directories "--directory=${CMAKE_BINARY_DIR}")
+    if(DEFINED arg_SOURCE_DIRECTORIES)
+        foreach(_src_dir ${arg_SOURCE_DIRECTORIES})
+            list(APPEND _coverage_directories "--directory=${_src_dir}")
+        endforeach()
+    else()
+        # Default to everything
+        list(APPEND _coverage_directories "--directory=${CMAKE_CURRENT_SOURCE_DIR}")
+    endif()
 
     # Setup target
-    add_custom_target(${_targetname}
+    add_custom_target(${arg_NAME}
 
         # Cleanup lcov
-        ${LCOV_EXECUTABLE} --no-external --gcov-tool ${GCOV_EXECUTABLE} --directory ${CMAKE_BINARY_DIR} --directory ${CMAKE_SOURCE_DIR}/components --zerocounters
+        ${LCOV_EXECUTABLE} --no-external --gcov-tool ${GCOV_EXECUTABLE} ${_coverage_directories} --zerocounters
 
-        # Run tests
-        COMMAND ${_testrunner} ${ARGV2}
+        # Run tests - allow for failing tests
+        COMMAND ${arg_RUNNER} || (exit 0)
 
         # Capture lcov counters and generating report
-        COMMAND ${LCOV_EXECUTABLE} --no-external --gcov-tool ${GCOV_EXECUTABLE} --directory ${CMAKE_BINARY_DIR} --directory ${CMAKE_SOURCE_DIR}/components --capture --output-file ${_targetname}.info
-        COMMAND ${LCOV_EXECUTABLE} --no-external --gcov-tool ${GCOV_EXECUTABLE} --directory ${CMAKE_BINARY_DIR} --directory ${CMAKE_SOURCE_DIR}/components --remove ${_targetname}.info '/usr/include/*' --output-file ${_targetname}.info.cleaned
-        COMMAND ${GENHTML_EXECUTABLE} -o ${_targetname} ${_targetname}.info.cleaned
-        COMMAND ${CMAKE_COMMAND} -E remove ${_targetname}.info ${_targetname}.info.cleaned
+        COMMAND ${LCOV_EXECUTABLE} --no-external --gcov-tool ${GCOV_EXECUTABLE} ${_coverage_directories} --capture --output-file ${arg_NAME}.info
```

Comment:
Have you tested that this works with more than one directory?

Reply:
Yes, the "default" is two - from my understanding it expects at least the binary directory and a source directory, so there should always be at least two.  CMake automatically expands the semicolon-delimited list into space-separated arguments in this case.